README.md
# Convert PerfProStudio files to ERG (Wahoo Kickr) files.

[![Travis CI Build Status][travis-badge]][travis-link]

  [travis-link]: https://travis-ci.org/plandes/clj-perferg
  [travis-badge]: https://travis-ci.org/plandes/clj-perferg.svg?branch=master

Convert PerfProStudio files to ERG mode files, which is used for trainers like
the WahooKickr.  I use [iMobileIntervals](http://imobileintervals.com), which
takes
[ERG](http://support.trainerroad.com/hc/en-us/articles/201869764-Erg-Mode-Explained) file
forms.  However, my coach exports files
as [PerfPro (.ppsmrx)](http://perfprostudio.com).  This utility converts to an
ERG file and also optionally provides and Excel summary file.


## Obtaining

The latest release binaries are
available [here](https://github.com/plandes/clj-perferg/releases/latest).

### Source

In your `project.clj` file, add:

[![Clojars Project](https://clojars.org/com.zensols.tools/erg/latest-version.svg)](https://clojars.org/com.zensols.tools/erg/)


## Usage

The app needs a `-p` for the `.ppsmrx` PerfPro file and
your [FTP](https://wattbike.com/us/functional-threshold-power).  You can
optionally provide `-e` and use `-h` to get a usage.

```bash
$ pp2erg -p Bike\ Climb\ 3\ \(4x4\)\ +\ \(4x1\)\ min.ppsmrx -f 244 -e
perferg: perfpro: processing Bike Climb 3 (4x4) + (4x1) min.ppsmrx with ftp 244
perferg: perfpro: wrote: Bike Climb 3 (4x4) + (4x1) min.erg
perferg: perfpro: wrote: Bike Climb 3 (4x4) + (4x1) min.xls
```

Usage:
```sql
$ pp2erg --help
usage: pp2erg [options]
Convert a perfpro .ppsmrx file to .erg and optionally .xls
  -l, --level <log level>     INFO  Log level to set in the Log4J2 system.
  -p, --perfpro <file>              The PerfPro .ppsmrx file input file
  -e, --excel                       If provided output an Excel summary file as well
  -f, --ftp <number>                The functional threshold power
  -i, --interpolate <number>  4.5   zone to percent FTP interplation constant
```

## Todo

* Currently the tool doesn't support individual zones--that's next.
* Add option to parse FTP description matches (i.e. *"seated climb 75% FTP"*
  gets an FTP of 75%.


## Documentation

API [documentation](https://plandes.github.io/clj-perferg/codox/index.html).


## Building

To build from source, do the folling:

- Install [Leiningen](http://leiningen.org) (this is just a script)
- Install [GNU make](https://www.gnu.org/software/make/)
- Install [Git](https://git-scm.com)
- Download the source: `git clone https://github.com/clj-perferg && cd clj-perferg`
- Download the make include files:
```bash
mkdir ../clj-zenbuild && wget -O - https://api.github.com/repos/plandes/clj-zenbuild/tarball | tar zxfv - -C ../clj-zenbuild --strip-components 1
```
- Build the distribution binaries: `make dist`

Note that you can also build a single jar file with all the dependencies with: `make uber`


## Changelog

An extensive changelog is available [here](CHANGELOG.md).


## License

Copyright © 2017 Paul Landes

Apache License version 2.0

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

[http://www.apache.org/licenses/LICENSE-2.0](http://www.apache.org/licenses/LICENSE-2.0)

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).


## [Unreleased]

## [0.0.2] - 2017-03-23
### Added
- Add FTP range support

### Changed
- Fix bug where only the first 10 intervals added to ERG.

## 0.0.1 - 2017-02-23
### Added
- Initial add converts perfpro files to ERG and XLS.

[Unreleased]: https://github.com/plandes/clj-perferg/compare/v0.0.2...HEAD
[0.0.2]: https://github.com/plandes/clj-perferg/compare/v0.0.1...v0.0.2


project.clj
(defproject com.zensols.tools/erg "0.1.0-SNAPSHOT"
  :description "Convert PerfProStudio files to ERG mode files, which is used for trainers like the WahooKickr."
  :url "https://github.com/plandes/clj-perferg"
  :license {:name "Apache License version 2.0"
            :url "https://www.apache.org/licenses/LICENSE-2.0"
            :distribution :repo}
  :plugins [[lein-codox "0.10.1"]
            [org.clojars.cvillecsteele/lein-git-version "1.2.7"]]
  :codox {:metadata {:doc/format :markdown}
          :project {:name "Convert PerfProStudio files to ERG mode files."}
          :output-path "target/doc/codox"
          :source-uri "https://github.com/plandes/clj-perferg/blob/v{version}/{filepath}#L{line}"}
  :git-version {:root-ns "zensols.erg"
                :path "src/clojure/zensols/erg"
                :version-cmd "git describe --match v*.* --abbrev=4 --dirty=-dirty"}
  :source-paths ["src/clojure"]
  :java-source-paths ["src/java"]
  :javac-options ["-Xlint:unchecked"]
  :jar-exclusions [#".gitignore"]
  :exclusions [org.slf4j/slf4j-log4j12
               log4j/log4j
               ch.qos.logback/logback-classic]
  :dependencies [[org.clojure/clojure "1.8.0"]

                 ;; logging for core
                 [org.apache.logging.log4j/log4j-core "2.7"]
                 [org.apache.logging.log4j/log4j-slf4j-impl "2.7"]
                 [org.apache.logging.log4j/log4j-1.2-api "2.7"]
                 [org.apache.logging.log4j/log4j-jcl "2.7"]

                 ;; json parse
                 [org.clojure/data.json "0.2.6"]

                 ;; spreadsheet
                 [com.zensols.tools/misc "0.0.4"]

                 ;; command line
                 [com.zensols.tools/actioncli "0.0.15"]]
  :pom-plugins [[org.codehaus.mojo/appassembler-maven-plugin "1.6"
                 {:configuration ([:programs
                                   [:program
                                    ([:mainClass "zensols.erg.core"]
                                     [:id "pp2erg"])]]
                                  [:environmentSetupFileName "setupenv"])}]]
  :profiles {:uberjar {:aot [zensols.erg.core]}
             :appassem {:aot :all}
             :snapshot {:git-version {:version-cmd "echo -snapshot"}}
             :dev
             {:jvm-opts ["-Dlog4j.configurationFile=test-resources/log4j2.xml" "-Xms4g" "-Xmx12g" "-XX:+UseConcMarkSweepGC"]
              :dependencies [[com.zensols/clj-append "1.0.5"]]}}
  :main zensols.erg.core)


src/clojure/zensols/erg/perfpro.clj
(ns zensols.erg.perfpro
  (:require [clojure.java.io :as io]
            [clojure.set :refer (rename-keys)]
            [clojure.data.json :as json]
            [clojure.string :as s]
            [clojure.tools.logging :as log])
  (:require [clj-excel.core :as excel])
  (:require [zensols.actioncli.log4j2 :as lu]
            [zensols.actioncli.parse :refer (with-exception)]
            [zensols.util.spreadsheet :as ss]))

(def ^:dynamic *default-percent-ftp-ranges*
  "Arary of arrays regular expressions of English sentence ranges."
  [#"(\d+)\s*(?:to|[\-])\s*(\d+)%"
   #"(\d+)%(?uix:ftp)"])

(def ^:dynamic *default-percent-ftp-descriptors*
  "Arary of arrays with each a regular expression of the description, a keyword
  identifier token and a percent to use for FTP."
  [[#"^(?ui)warm\s*up$" :warmup 0.4]
   [#"^(?ui)cool\s*down$" :sprint 0.3]
   [#"^(?ui)(easy\s*)?recovery?$" :recovery 0.6]
   [#"^(?ui)fast\s*spin$" :fast-spin 0.3]
   [#"^(?ui)easy\s*spin$" :fast-spin 0.35]
   [#"^(?ui)spin$" :spin 0.4]
   [#"^(?ui)sprint$" :sprint 0.9]])

(def ^:dynamic *linear-interpolate-ftp-zone* 4.5)

(defn- parse-ppsmrx
  "Parse a .ppsmrx JSON file."
  [file]
  (log/debugf "parsing perfpro file: %s" file)
  (with-open [reader (io/reader file)]
    (-> (json/read-json reader)
        (select-keys [:set_fields :workoutType :sets :comments :name])
        (rename-keys {:set_fields :fields
                      :workoutType :type}))))

(defn- description-to-ftp-range
  [s]
  (->> *default-percent-ftp-ranges*
       (some (fn [pat]
               (let [[_ st en] (re-find pat s)]
                 (cond (and st en) (map read-string [st en])
                       st (map read-string [st st])))))))

(defn- description-to-ftp-descriptor
  "Guess the percent FTP by the description.  This is given by "
  [s]
  (->> *default-percent-ftp-descriptors*
       (some (fn [[pat desc percent]]
               (if (re-find pat s)
                 {:descriptor desc
                  :percent percent})))
       ((fn [data]
          (or data (-> (format "No known descriptor for: <%s>" s)
                       (ex-info {:descriptor s})
                       throw))))))

(defn- zone-to-percent
  "Convert from an FTP zone to an FTP percent.  **desc** is the description of
  the interval and num is the `start` field."
  [desc num]
  (log/debugf "zone to percent: %s, %d" desc num)
  (if-let [ftp-range (description-to-ftp-range desc)]
    (let [high (apply max ftp-range)
          low (apply min ftp-range)]
      (/ (+ (/ (- high low) 2.0) low) 100.0))
    (let [fdesc (description-to-ftp-descriptor desc)]
      (if (and (< 0 num) (> 7 num))
        (/ num *linear-interpolate-ftp-zone*)
        (or (-> fdesc :percent)
            (-> (format "Don't know how to convert percent '%s' of %d" desc num)
                (ex-info {:desc desc :num num})
                throw))))))

(defn- power-sets
  "Create power sets used in the ERG.  **ftp** is the athlete's FTP and
  **ppdata** is parsed by [[parse-ppsmrx]]."
  [ftp ppdata]
  (log/debugf "computing power sets on ftp=%d" ftp)
  (let [{:keys [fields sets]} ppdata
        fields (map keyword fields)]
    (->> sets
         (map (fn [row]
                (zipmap fields row)))
         (map (fn [{:keys [start seconds description mode targetcad] :as m}]
                (when (> (-> description s/trim count) 0)
                  (log/debugf "[%d, %d]: <%s>, m=%s, tc=%d"
                              start seconds description mode targetcad)
                  (let [minutes (/ (double seconds) 60.0)
                        percent (if (= mode "M")
                                  (-> (description-to-ftp-descriptor description)
                                      :percent)
                                  (zone-to-percent description start))
                        watts (* percent ftp)]
                    {:description description
                     :cadence targetcad
                     :seconds seconds
                     :minutes minutes
                     :percent percent
                     :watts watts}))))
         (remove nil?))))

(defn- save-power-sets
  "Save an Excel file using power sets (**sets**).  **fields** are the keys
  returned in maps from [[power-sets]] and **file** is the output file."
  [fields tab-name file sets]
  (log/debugf "saving excel file: %s, %s, %s" file fields tab-name)
  (-> (excel/build-workbook
       (excel/workbook-hssf)
       {tab-name
        (->> sets
             (cons fields)
             ss/headerize)})
      (ss/autosize-columns)
      (excel/save file)))

(defn- save-as-excel
  "Massage data and save with [[save-power-sets]]."
  [ppdata power-sets out-file]
  (let [fields [:description :minutes :percent :watts :cadence]]
    (->> power-sets
         (map (fn [row]
                (map #(get row %) fields)))
         (save-power-sets (map #(-> % name s/capitalize) fields) (:name ppdata)
                          out-file))))

(defn- course-data
  "Create the ERG *course* data."
  [power-sets]
  (->> power-sets
       (reduce (fn [{:keys [rows time description]} {:keys [minutes watts]}]
                 (let [start time
                       end (+ time minutes)
                       mmap {:watts watts
                             :description description}]
                   {:time end
                    :rows (concat rows [(assoc mmap :time start)
                                        (assoc mmap :time end)])}))
               {:rows []
                :time 0})
       :rows))

(defn- compose-erg
  "Print the ERG file to standard out."
  [ftp out-file power-sets ppdata]
  (let [ftp 244
        {:keys [name type]} ppdata
        course-data (course-data power-sets)]
    (println "[COURSE HEADER]")
    (->> (format "MINUTES WATTS
VERSION = 2
UNITS = ENGLISH
DESCRIPTION = %s (type: %s)
FILE NAME = %s
FTP = %d
MINUTES WATTS
[END COURSE HEADER]
[COURSE DATA]"
                 name type
                 out-file
                 ftp)
         println)
    (->> course-data
         (map (fn [{:keys [time watts]}]
                (format "%.2f\t%d" (double time) (int watts))))
         (map println)
         doall)
    (println "[END COURSE DATA]")))

(defn- file-to-name
  "Get the name portion of a file."
  [file]
  (->> file
       .getName
       (#(or (-> (re-find #"^(.*)\." %) second) %))))

(defn read-and-save
  "Read and parse the PerfPro `.ppsmrx` file, then output the ERG file.

* **ftp** is the athlete's FTP (see project docs)
* **perf-pro-file** is the `.ppsrmx` file
* **export-excel?** if non-nil create an Excel summary file as well"
  [ftp perf-pro-file export-excel?]
  (let [in-file perf-pro-file
        file-name (file-to-name in-file)
        parent-file (.getParentFile in-file)
        erg-file (->> file-name (format "%s.erg") (io/file parent-file))
        xls-file (->> file-name (format "%s.xls") (io/file parent-file))
        ppdata (parse-ppsmrx in-file)
        power-sets (power-sets ftp ppdata)]
    (with-open [writer (io/writer erg-file)]
      (binding [*out* writer]
        (compose-erg ftp erg-file power-sets ppdata)))
    (log/infof "wrote: %s" erg-file)
    (when export-excel?
      (save-as-excel ppdata power-sets xls-file)
      (log/infof "wrote: %s" xls-file))))

(def convert-command
  "CLI command to convert PerfPro .ppsmrx files."
  {:description "Convert a perfpro .ppsmrx file to .erg and optionally .xls"
   :options
   [(lu/log-level-set-option)
    ["-p" "--perfpro" "The PerfPro .ppsmrx file input file"
     :required "<file>"
     :parse-fn io/file]
    ["-e" "--excel" "If provided output an Excel summary file as well"]
    ["-f" "--ftp" "The functional threshold power"
     :required "<number>"
     :parse-fn read-string
     :validate [#(> % 50) "Must be a positive number above 50"]]
    ["-i" "--interpolate" "zone to percent FTP interplation constant"
     :required "<number>"
     :parse-fn read-string
     :default *linear-interpolate-ftp-zone*
     :validate [#(> % 1) "Must be a positive number above 1"]]]
   :app (fn [{:keys [perfpro ftp interpolate excel] :as opts} & args]
          (with-exception
            (binding [*linear-interpolate-ftp-zone* interpolate]
              (if (nil? perfpro) (throw (ex-info "Missing -p option" {})))
              (if (nil? ftp) (throw (ex-info "Missing -f option" {})))
              (log/infof "processing %s with ftp %s" perfpro ftp)
              (log/debugf "excel: %s" excel)
              (read-and-save ftp perfpro excel))))})
